### ISSUE TITLE
Use std::terminate_handler

### ISSUE THREAD

**Issue by takev** [hikogui/hikogui#122]
Certain parts of the application uses tt_debugger_abort() to terminate the application.
Instead set a std::terminate_handler. Which:

- Dumps a stacktrace to the log file. (not part of this issue)
- Dumps a mini core. (not part of this issue) 
- Breaks the debugger, or
- Shows a dialogue to the user with the reason of the "crash".

The optional error message should be written to a thread_local variable that is picked up by the terminate handler.

std::current_exception() can I think be used to catch the unhandled C++ exception.

Also make a unhandled exception handler for structured exceptions and rethrow it as a normal exception.

**Comment by jhalakpatel** [comment]
@takev can I work on this issue?

**Comment by jhalakpatel** [comment]
If that is OK with you can you elaborate on the task. IIUC, we need to implement something like this:

Implement `tt_terminate_handler` and replace `tt_debugger_abort` with it.

`
void tt_terminate_handler()
{
    // create crash message similar to tt_debugger_abort
   
    std::terminate_handler th = [&]()->void
    {
         // debugger break
        // handle unhandled exceptions using std::current_exception
        // unhandled exception handler for structured exceptions and throw as a normal exception     
    }
    
    std::abort();
}
`

Please review and let me if I understood the issue correctly. If not, please elaborate and help me understand. Apologies in advance as I have not return lot of exception handling code so far, but want to learn from this opportunity.

**Comment by takev** [comment]
All calls to `tt_debugger_abort()` should be replaced with `std::terminate()`.

`std::terminate()` will call a registered terminate handler. You can register your own terminate handler by passing the function pointer to `std::set_terminate()`. This function will return a pointer to the previous function, which our own terminate handler should call once it is finished, this is either a pointer to `std::abort` or another in-between terminate handler registered by the application or another library

And our terminate handler should just be a normal function instead of a lambda. Call it
```
inline void terminate_handler() noexcept
{
}
```
in terminate_handler.hpp.

Registering the terminate handler should be done in crt.hpp.

**Comment by jhalakpatel** [comment]
Thanks, @takev. 
I understand that `std::set_terminate` would return a pointer to the previous function (i.e. previous terminate handler if registered by some other library or application), how should our terminal handler capture that?
Also, IIUC `std::set_terminate` needs to be registered before any exception is thrown. As you recommended to add that in crt.hpp, should this be registered after `start_system()` (as noexcept) and before `tt_main` ? (I was also wondering if we need to register terminate handler for each thread or just the main thread?)
Thanks for your time.

**Comment by takev** [comment]
You store the return value of set_terminate in a global variable in the terminate_handler.hpp.

set_terminate itself uses a global variable so you only need to do this once, and it will work for all threads.

I cannot find the requirement that it needs to be registered before and exception is thrown. But between start_system and tt_main seems like a good place.
